enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,444 @@
+proposal, since enabling swap on control-plane nodes is not supported.
+
+### Implementation Details/Notes/Constraints
+
+The design is driven by the following guiding principles:
+
+* System services are more important than workloads. Because workload
+  health depends on the health of system services.
+* Try to stay aligned to upstream Kubernetes swap behavior in order
+  to ease the transition to Kubernetes SWAP once available
+
+#### WASP Agent
+
+Overall the agent is intended to align - and specifically not
+conflict - with the upstream [Kubernetes SWAP design](https://kubernetes.io/blog/2023/08/24/swap-linux-beta/) and behavior in
+order to simplify a transition. One extra functionality that the agent
+currently provides is the swap-based evictions. 
+
+The agent uses [OCI Hook] to enable swap via container cgroup. The hook script 
+sets the containers cgroup `memory.swap.max=max`.
+
+* **Tech Preview**
+  * Limited to burstable QoS class pods.
+  * Uses `UnlimitedSwap` approach. Using OCI hook each container is started with `memory.swap.max=max` in its cgroup.
+* **General Availability**
+  * Container starts with `UnlimitedSwap` as in TP, then reconciled to `LimitedSwap` by the wasp-agent limited swap controller.
+    * For more info, refer to the upstream documentation on how to calculate [limited swap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#steps-to-calculate-swap-limit). 
+  * Added swap-based evictions.
+  * Added critical workload protection (same as kube swap).
+
+![wasp](https://github.com/user-attachments/assets/314d387d-788a-4aa6-9275-c882f5f329aa)
+
+**NOTE**: In WASP GA setting the swap limit isn't atomic, it can happen that a container application will start
+to allocate memory before the limit is set by the wasp-agent.
+
+#### Swap-based evictions 
+
+Dealing with memory pressure on a node is differentiating the TP fom GA.
+
+** **Technology Preview** - `memory.high` is set on the `kubepods.slice`
+in order to force the node to swap, once the `kubepods.slice` memory is
+filling up. Only once swap is full, the system will cross `memory.high`
+and trigger soft evictions.
+
+* Pro
+    * Simple to achieve.
+* Con
+    * A lot of memory pressure has to be present in order to trigger
+      soft eviction.
+    * Once `memory.high` is reached, the whole `kubepods.slice` is throttled
+      and cannot allocate memory, which might lead to applications crashing.
+
+** **General Availability** - Memory-based soft eviction is going to
+be disabled, in favor of enabling swap-based hard evictions, based on new
+swap traffic and swap utilization eviction metrics.
+
+* Pro
+    * Eviction on the basis of swap pressure, not only memory pressure.
+    * [LLN] applies, because all pods share the nodes memory
+* Con
+    * Swap-based evictions are made through a 3rd party container, which means
+      it has to be done through an API-initiated eviction.
+
+
+#### Machine-level swap configuration
```

Comment:
can you explicitly  state the intention is to have the swap operator set up these things?

Reply:
Done.